feat: add SSL certificate validation for Druid

villebro opened the pull request:
### CATEGORY

Choose one

- [ ] Bug Fix
- [x] Enhancement (new features, refinement)
- [ ] Refactor
- [x] Add tests
- [ ] Build / Development Environment
- [x] Documentation

### SUMMARY
This adds support for specifying SSL certificates on a per database basis.  This is done by adding a new field `server_cert` to the `dbs` table and using that in the `pydruid.connect()` call. Unfortunately Python doesn't yet support passing certificates either as strings or file-like objects (see the [PR](https://bugs.python.org/issue16487) that has been 8 years in the making and the stalled [PEP-543](https://www.python.org/dev/peps/pep-0543/)), hence the certificate has to be persisted to disk ☹️ To avoid rewriting the certificate to disk every time it is used, the file path is made deterministic by naming it with the md5 of the certificate contents and only written to disk if it doesn't yet exist in the designated path. Unit tests have also been added to verify that only valid certificates are persisted to disk.

### Screenshot
![image](https://user-images.githubusercontent.com/33317356/77643753-97a9f000-6f68-11ea-81f6-da6722571fb7.png)

### TEST PLAN
New unit tests + local testing

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [x] Changes UI
- [x] Requires DB Migration.
- [x] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS
@mistercrunch @john-bodley @dpgaspar @willbarrett

villebro committed:
feat: add SSL certificate feature

dpgaspar commented on tests/utils_tests.py:
```diff
@@ -1221,3 +1222,33 @@ def test_build_extra_filters(self):
+XeA3AKTX/OdYWJvr5YIgeQ==
+-----END CERTIFICATE-----"""
+        path = create_temporary_ssl_cert_file(valid_certificate)
+        self.assertIn("5b0df668ac310be3f15c489f303d7d01", path)
```
Small suggestion, we could calculate the cert's MD5 here instead of relying on the hardcoded value. Would it make sense to place `valid_certificate` on the fixtures test folder?

villebro replied:
Good idea, will change it.

mistercrunch replied:
NIT: something like this could go in a `fixtures.py` module

dpgaspar commented on superset/utils/core.py:
```diff
@@ -1163,6 +1166,26 @@ def get_username() -> Optional[str]:
+    path = os.path.join(tempfile.gettempdir(), filename)
+    if not os.path.exists(path):
+        # Validate certificate prior to persisting to temporary directory
+        x509.load_pem_x509_certificate(certificate.encode("utf-8"), default_backend())
```
Looking at the tests, I'm guessing this raises `ValueError` we could wrap and re-raise with a custom exception. Should `get_sqla_engine` catch these exceptions how are all these handled?

villebro replied:
I'm open to suggestions on naming. `ValidationException`? `CertificateValidationException`?

villebro commented on superset/migrations/versions/b5998378c225_add_certificate_to_dbs.py:
```diff
@@ -0,0 +1,49 @@
+        kwargs["postgresql_using"] = "encrypted_extra::bytea"
+    op.add_column(
+        "dbs",
+        sa.Column("server_cert", EncryptedType(sa.Text()), nullable=True, **kwargs),
```
I did but took it out, as `pydruid` doesn't support it yet. However, I opened a [PR](https://github.com/druid-io/pydruid/pull/201) for that, and will add the client cert field once that PR is merged.

villebro commented on superset/db_engine_specs/druid.py:
```diff
@@ -47,3 +49,18 @@ class DruidEngineSpec(BaseEngineSpec):  # pylint: disable=abstract-method
+        if database.server_cert:
+            connect_args["scheme"] = "https"
+            path = utils.create_temporary_ssl_cert_file(database.server_cert)
+            connect_args["ssl_verify_cert"] = path
```
Good idea. This raises `ValueError` if it doesn't validate, and it could be really helpful to have a specific `Exception` for that case, e.g. in the `Test Connection` where I suspect most people will hit errors.

villebro commented on superset/views/database/mixins.py:
```diff
@@ -196,6 +204,9 @@ def _pre_add_update(self, database):
         self.check_encrypted_extra(database)
+        database.server_cert = (
+            database.server_cert.strip() if database.server_cert else ""
+        )
```
Good idea, I'll break it out into a separate util function.

dpgaspar replied:
"server_cert" is not being validated before persisting to the DB, are we relying on `testconn`. It would be nice to add `parse_ssl_cert` to the pre_add, pre_update hooks

craig-rueda commented on superset/utils/core.py:
```diff
@@ -1163,6 +1166,26 @@ def get_username() -> Optional[str]:
+    :return: The path to the certificate file
+    """
+    filename = hashlib.md5(certificate.encode("utf-8")).hexdigest()
+    path = os.path.join(tempfile.gettempdir(), filename)
```
Can we make this dir configurable? Placing it in temp works, but users might want to be able to tune this. Maybe default to  temp?

villebro replied:
Good idea, will add a config param.

john-bodley commented on superset/utils/core.py:
```diff
@@ -1163,6 +1166,26 @@ def get_username() -> Optional[str]:
         return None
 
 
+def create_temporary_ssl_cert_file(certificate: str) -> str:
```
This doesn’t feel temporary to me. Maybe the method and docstring should be updated.

villebro replied:
Good point. This was a remnant from when I thought these would really be temporary. I will update accordingly.

john-bodley commented on superset/db_engine_specs/base.py:
```diff
@@ -959,3 +959,15 @@ def mutate_db_for_connection_test(database: "Database") -> None:
+        Some databases require passing additional non-standard parameters to database
+        connections, for example client certificates.
+
+        :param database: instance to be mutated
```
Nit. Missing `connect_args` params.

mistercrunch replied:
Wait, connect_args is an attribute of the database isn't it?

villebro replied:
Yes, or more specifically it's in `database.extra`, which is in JSON format. The reason it's being passed as a separate argument is that it's already been parsed out of `extra`. We could optionally inject this into `extra`, too, but that might be more confusing, as then we'd have to `json.parse`, mutate and `json.dumps` again.

villebro replied:
@mistercrunch I changed this so extras is parsed in `db_engine_specs`, so that we don't need to mutate the `Database` instance (this caused the mutations to get persisted in the db when saving the database).

john-bodley commented on superset/db_engine_specs/druid.py:
```diff
@@ -47,3 +49,18 @@ class DruidEngineSpec(BaseEngineSpec):  # pylint: disable=abstract-method
+        Some databases require passing additional non-standard parameters to database
+        connections, for example client certificates.
+
+        :param database: instance to be mutated
```
See previous comment.

villebro committed:
Address comments

villebro committed:
don't mutate extras

villebro commented:
All comments should now be addressed.

dpgaspar commented on superset/db_engine_specs/base.py:
```diff
@@ -959,3 +963,20 @@ def mutate_db_for_connection_test(database: "Database") -> None:
+        like passing certificates to `extra`. This can be done here.
+
+        :param database: database instance from which to extract extras
+        """
```
nit: it's nice to add a `:raises:` on the docstring when a method can raise an exception

villebro replied:
Ah, missed that one, will update.

dpgaspar commented on superset/db_engine_specs/druid.py:
```diff
@@ -47,3 +52,26 @@ class DruidEngineSpec(BaseEngineSpec):  # pylint: disable=abstract-method
+        """
+        For Druid, the path to a SSL certificate is placed in `connect_args`.
+
+        :param database: database instance from which to extract extras
```
nit: it's nice to add a :raises: on the docstring when a method can raise an exception

villebro committed:
Address comments and add polish

villebro commented:
Addressed overlooked comment by @dpgaspar and added error message when certificate is invalid:
![image](https://user-images.githubusercontent.com/33317356/77700643-3d874a00-6fbd-11ea-8e28-fd932a6e7267.png)

dpgaspar reviewed (approved):
Looks good!

villebro committed:
Add further polish

brachipa commented:
Hi, 
I know that this PR is already closed, but may I just raise issue if I can?

I can't configure db connection with cert, it gave me exception: 
```
ERROR:superset.views.core:Unexpected error connect() got an unexpected keyword argument 'ssl_verify_cert'
```
I see in your code that you set in this field, the cert path. and I print the path into the logs and see the correct file name,I also check, and this file is saved in my file system.

(also setting this, doesn't really help, it gives same error, seems like you ignore that property
`"engine_params": {"connect_args": {"scheme": "https", "ssl_verify_cert": false}}`

So why does it fail? seems like something sqlalchemy.
Any idea?
